File: cpp/Balance_a_Binary_Search_Tree_1382.cpp
class Solution {
public:

    void storeInOrder(TreeNode* root, vector<TreeNode*>& nodes) {
        if (!root) return;
        storeInOrder(root->left, nodes);
        nodes.push_back(root);
        storeInOrder(root->right, nodes);
    }

    TreeNode* buildBST(vector <TreeNode*>& nodes, int start, int end) {
        if (start > end) return nullptr;
        int mid = start + (end - start) / 2;
        TreeNode* root = nodes[mid];
        root->left = buildBST(nodes, start, mid - 1);
        root->right = buildBST(nodes, mid + 1, end);
        return root;
    }

    TreeNode* balanceBST(TreeNode* root) {
        vector <TreeNode*> nodes;
        storeInOrder(root, nodes);
        return buildBST(nodes, 0, nodes.size() - 1);
    }
};

File: python/Balance_a_Binary_Search_Tree_1382.py
class Solution:

    def storeInOrder(self, root, nodes):
        if not root:
            return
        self.storeInOrder(root.left, nodes)
        nodes.append(root)
        self.storeInOrder(root.right, nodes)

    def buildBST(self, nodes, start, end):
        if start > end:
            return None
        mid = (start + end) // 2
        root = nodes[mid]
        root.left = self.buildBST(nodes, start, mid - 1)
        root.right = self.buildBST(nodes, mid + 1, end)
        return root

    def balanceBST(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        nodes = []
        self.storeInOrder(root, nodes)
        return self.buildBST(nodes, 0, len(nodes) - 1)
